aten/src/TH/generic/THTensor.cpp

```diff
@@ -766,6 +781,82 @@ void THTensor_(setStorageNd)(THTensor *self, THStorage *storage, ptrdiff_t stora
   THTensor_(resizeNdLegacy)(self, nDimension, size, stride);
 }
 
+
+void THTensor_(resizeNd)(THTensor *self, int nDimension, int64_t *size, int64_t *stride)
+{
+  int d;
+  ptrdiff_t totalSize;
+  bool hascorrectsize = true;
+
+#ifndef USE_TH_SCALAR
+  AT_CHECK(nDimension > 0, "resizeNd nDimension must be greater than 0");
+#else
+  AT_CHECK(nDimension >= 0, "resizeNd nDimension must be non-negative");
+#endif
+
+  for(d = 0; d < nDimension; d++)
+  {
+#ifndef USE_TH_SIZE_ZERO_DIM
+    // we can't support this unless we have arbitrary 0-sized dimensions, but some calls to this
+    // currently exist and expect a size [0] tensor to be returned.
+    if (d == 0 && size[d] == 0) {
+      nDimension = 1;
+      break;
+    }
+    AT_CHECK(size[d] > 0, "sizes must be non-negative");
+#endif
+    if((self->dim() > d) && (size[d] != self->size[d])) {
+      hascorrectsize = false;
+    }
+
+    // NB: this used to test that stride[d] was >= 0
+    if((self->dim() > d) && stride && (stride[d] != self->stride[d])) {
+      hascorrectsize = false;
+    }
+  }
+
+  if(nDimension != self->dim()) {
+    hascorrectsize = false;
+  }
+
+  if(hascorrectsize) {
+    return;
+  }
+
+  if(nDimension != self->dim())
+  {
+    self->size = (int64_t *)THRealloc(self->size, sizeof(int64_t)*nDimension);
+    self->stride = (int64_t *)THRealloc(self->stride, sizeof(int64_t)*nDimension);
+    self->dim_ = nDimension;
+  }
+
+  totalSize = 1;
+  for(d = nDimension-1; d >= 0; d--)
```

Comment:
Did you want to write this code to also support negative strides? I worked out the math and wrote this helper function: https://github.com/ezyang/pytorch/blob/c10/caffe2/c10_prototype/c10/Utils.h#L30